text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,233 @@
+
+The second user group interacts with the described feature through attributes. These attributes allow them to hint the compiler to emit specific error messages in certain cases. The `#[diagnostic]` attribute namespace provides a general framework for what can and can't be done by such an attribute. As of this users won't interact directly with the attribute namespace itself. 
+
+The `#[diagnostic::on_unimplemented]` attribute allows to hint the compiler to emit a specific error message if a certain trait is not implemented. This attribute should provide the following interface:
+
+```rust
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="note"
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+
+
+fn iterate_chars<I: MyIterator<char>>(i: I) {
+    // ...
+}
+fn main() {
+    iterate_chars(&[1, 2, 3][..]);
+}
+```
+
+which might result in the compiler emitting the following error message:
+
+```
+error[E0277]: message
+  --> <anon>:14:5
+   |
+14 |     iterate_chars(&[1, 2, 3][..]);
+   |     ^^^^^^^^^^^^^ label
+   |
+   = note: note
+   = help: the trait `MyIterator<char>` is not implemented for `&[{integer}]`
+   = note: required by `iterate_chars`
+```
+
+I expect the new attributes to be documented on the existing [Diagnostics](https://doc.rust-lang.org/reference/attributes/diagnostics.html) attributes page on the rust reference similar to existing attributes like for example `#[deprecated]`
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The `#[diagnostic]` attribute namespace
+
+This RFC proposes to introduce a new built-in `#[diagnostic]` tool attribute namespace. This namespace is supposed to contain different attributes, which allow users to hint the compiler to emit specific diagnostic messages in certain cases like type mismatches, unsatisfied trait bounds or similar situations. By collecting such attributes in a common namespace it is easier for users to find useful attributes and it is easier for the language team to establish a set of common rules for these attributes. This opens the possibility to delegate the design of specific attributes to other teams later on.
+
+Attributes in this namespace are generally expected to be formed like:
+```rust
+#[diagnostic::attribute(option)]
+```
+where several `option` entries can appear in the same attribute. `option` is expected to be a valid attribute argument in this position. 
+
+Any attribute in this namespace may:
+
+* Hint the compiler to emit a specific diagnostic message in a specific situation
+* Only affect the messages emitted by a compiler 
+
+Any attribute in this namespace is not allowed to:
+
+* Change the result of the compilation, which means applying such an attribute should never cause a compilation error as long as they are syntactically valid
```

Comment:
Should this say “applying _or removing_ such an attribute should never cause a compilation error”?

Reply:
I would argue that this is already somewhat implied by the first part of the sentence (cannot change the result of the compilation). That written, it might be meaningful to add it explicitly to the text as well.